cowprotocol/cowswap
# [Widget enhancement] 17 - Emit toast for posting orders

# Summary

Part of the series https://github.com/cowprotocol/cowswap/pull/3812

This PR adds the first toast message: `Posting orders`. I had to make quite a big change because I needed the middleware to be stateless, and some relevant information was missing from the redux actions. 

## Details
To make the toast messages simpler for the clients, it will send a text message (as a `string`) so it can be passed to their message/toast library. 

This introduces a small problem, because we already have our toast messages, and we format using react components, if we want a string message version of the same, we will need to duplicate some logic. 

This PR tries to implement this first toast message, and for that it will need to:
- [x] Refactor the code to format token symbols/names (we don't use the `TokenSymbol` component, but it will use a similar formatting logic). Where possible, this PR refactors both codes
- [x] Refactor the code to format token amounts with currency (we don't use the `TokenAmount` component, but it will use a similar formatting logic). Where possible, this PR refactors both codes
- [x] Implement `getPendingOrderNotificationToast` to get the equivalent string version for the toast of `PendingOrderNotification` component
- [x] Implement `formatTokenAmountWithSymbol` to get the equivalent string version for the toast of `Token` component
- [x] Make sure the redux action payloads are self contained, so they don't rely on hooks or any other info later. This is good for some reasons:
    - Become state-less, for example, if we get a message for an order created by a safe (`isSafeWallet = true` ), and somehow we change the wallet to one that is not a safe (`isSafeWallet = true`), if we mix data from the payload from data from hooks then we can get into trouble
    - This was needed also, for implementing the message in a pure function that wouldn't depend on hooks


Additionally, this PR deletes some dead code and testing log messages


<img width="1209" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/9dc71b98-b716-42af-9903-91a745030ccc">


## Cosmos
Because now `PendingOrderNotification` is a dumb container and it needs to receive the `isSafeWallet` flag, I added the safe variations of the fixtures.

EOA:
<img width="1183" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/c44c7ab0-4cf4-485b-8d03-ce20ebeded68">

Safe version
<img width="1183" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/cbd74b13-1a43-4d94-82bf-e247f65b6f3e">

## Unit tests
Additionally adds some unit tests

<img width="911" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/4fba0df6-bc31-4806-a5f5-9cd8095b70ed">


## How to tests
1. Open the configurator: https://widget-configurator-git-widget-enhancement-event-99b848-cowswap.vercel.app
     - EDIT: You will need to test locally the configurator... see problems with the configurator link) 
3. Connect in Sepolia (or any network)
4. Show dev tools, and filter console events by `[configurator:`
5. Place an order
6. Check the console messages (should show something like the screenshot above)

## Issue with the links of the PREVIEW
I saw something is not working with the preview links in the configurator.

Is not part of this PR, so please test locally.

I suspect it has to do with the slash of the branch name.

Basically the good URL don't matches the one the configurator creates:
* THE ONE THIS PR CREATES: https://swap-dev-git-widget-enhancement-event-99b848-cowswap.vercel.app
* GOOD: https://swap-dev-git-widget-enhancement-events-and-params-17-cowswap.vercel.app

I checked and it has to do with the fact we assume the URL of the widget and cowswap generated by Vercel follows the same patter, but for some reason, in this PR doesnt
-  see https://github.com/cowprotocol/cowswap/blob/eff8657b05adc54b1d13eb58c08578cec1ebc45c/libs/widget-lib/src/consts.ts#L8
